### PR TITLE
[Enhancement] port `ORCHdfsFileStream` to `OrcScanner`

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [X] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #https://github.com/StarRocks/starrocks/issues/11329

## Problem Summary(Required) ：

Class `ORCHdfsFileStream` has capability to do pre-read and io coalesce functionalities. However `OrcScanner` does not use it, which makes broker load extremly slow and do a lot of random IO.

With this enhancement, case2 can be cutdown from 35s -> 5/6s.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: be/src/formats/orc/orc_chunk_reader.cpp [hunk 1]
```diff
@@ -2281,4 +2282,91 @@ int OrcChunkReader::get_column_id_by_name(const std::string& name) const {
...
+        throw orc::ParseError("Buffer is null");
+    }
+    Status status;
+    if (direct || !_buffer_stream_enabled) {
+        status = _file->read_at_fully(offset, buf, length);
+    } else {
+        const uint8_t* ptr = nullptr;
+        size_t nbytes = length;
+        status = _buffer_stream.get_bytes(&ptr, offset, &nbytes, false);
+        DCHECK_EQ(nbytes, length);
```
Comment (xiaoyong-z) [inline be/src/formats/orc/orc_chunk_reader.cpp:2342]: will crash here if there exists some problem with the network

**Comment by xiaoyong-z** [comment]
would be better if add all these tests:
<div dir="auto" style="box-sizing: border-box; color: rgb(201, 209, 217); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">case1 : 1 file of 20,000 rows, with 590 columns, all columns are string and decimal</div><div dir="auto" style="box-sizing: border-box; color: rgb(201, 209, 217); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">case2: 20 files with 1000 rows each. with 590 columns in total, all columns are string and decimal</div><div dir="auto" style="box-sizing: border-box; color: rgb(201, 209, 217); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">case3: 1 file of 20,000 rows, with 590 columns, all columns are int</div><div dir="auto" style="box-sizing: border-box; color: rgb(201, 209, 217); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">case4: 100 files with 1000 rows each, with 590 columns in total, all columns are int</div><div dir="auto" style="box-sizing: border-box; color: rgb(201, 209, 217); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">case5: 1 file with 400000 rows, with 30 columns in total, all columns are string</div><div dir="auto" style="box-sizing: border-box; color: rgb(201, 209, 217); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px;"></p>

  | CSV | Parquet | ORC | ORC after optimization
-- | -- | -- | -- | --
case1 | 4s | 6s | 14s | 
case2 | 6s | 6s | 35s | 
case3 | 5s | 4s | 2s | 
case4 | 7s | 8s | 6s | 
case5 | 7s | 6s | 7s | 

</div>

File: be/src/formats/orc/orc_chunk_reader.cpp (hunk 1 above)
Comment (dirtysalt) [reply be/src/formats/orc/orc_chunk_reader.cpp:2342]: in `get_bytes`, it calles `read_at_fully`, which expects to `nbytes`.

```
Status SharedBufferedInputStream::get_bytes(const uint8_t** buffer, size_t offset, size_t* nbytes, bool peek) {
    auto iter = _map.upper_bound(offset);
    if (iter == _map.end()) {
        return Status::RuntimeError("failed to find shared buffer based on offset");
    }
    SharedBuffer& sb = iter->second;
    if ((sb.offset > offset) || (sb.offset + sb.size) < (offset + *nbytes)) {
        return Status::RuntimeError("bad construction of shared buffer");
    }
    if (sb.buffer.capacity() == 0) {
        sb.buffer.reserve(sb.size);
        RETURN_IF_ERROR(_file->read_at_fully(sb.offset, sb.buffer.data(), sb.size));
    }
    *buffer = sb.buffer.data() + offset - sb.offset;
    return Status::OK();
}
```

File: be/src/formats/orc/orc_chunk_reader.cpp [hunk 2]
```diff
@@ -2281,4 +2282,91 @@ int OrcChunkReader::get_column_id_by_name(const std::string& name) const {
...
+    if (length > cache_max_size) return;
+    if (canUseCacheBuffer(offset, length)) return;
+
+    // If this stripe is small, probably other stripes are also small
+    // we combine those reads into one, and try to read several stripes in one shot.
+    if (scope == orc::InputStream::PrepareCacheScope::READ_FULL_STRIPE) {
+        length = std::min(_length - offset, cache_max_size);
+    }
+
+    _cache_buffer.resize(length);
```
Comment (xiaoyong-z) [inline be/src/formats/orc/orc_chunk_reader.cpp:2305]: reserve?

File: be/src/formats/orc/orc_chunk_reader.cpp [hunk 3]
```diff
@@ -2281,4 +2282,91 @@ int OrcChunkReader::get_column_id_by_name(const std::string& name) const {
...
+    Status status;
+    if (direct || !_buffer_stream_enabled) {
+        status = _file->read_at_fully(offset, buf, length);
+    } else {
+        const uint8_t* ptr = nullptr;
+        size_t nbytes = length;
+        status = _buffer_stream.get_bytes(&ptr, offset, &nbytes, false);
+        DCHECK_EQ(nbytes, length);
+        if (status.ok()) {
+            ::memcpy(buf, ptr, length);
```
Comment (xiaoyong-z) [inline be/src/formats/orc/orc_chunk_reader.cpp:2344]: we can let the buf directly pointer to ptr，thus can reduce the memory copy?

File: be/src/formats/orc/orc_chunk_reader.cpp [hunk 4]
```diff
@@ -2281,4 +2282,91 @@ int OrcChunkReader::get_column_id_by_name(const std::string& name) const {
...
+
+    // If this stripe is small, probably other stripes are also small
+    // we combine those reads into one, and try to read several stripes in one shot.
+    if (scope == orc::InputStream::PrepareCacheScope::READ_FULL_STRIPE) {
+        length = std::min(_length - offset, cache_max_size);
+    }
+
+    _cache_buffer.resize(length);
+    _cache_offset = offset;
+    doRead(_cache_buffer.data(), length, offset, true);
```
Comment (xiaoyong-z) [inline be/src/formats/orc/orc_chunk_reader.cpp:2307]: // If this stripe is small, probably other stripes are also small
// we combine those reads into one, and try to read several stripes in one shot.
if there are many small strips, why we alaways read from the beginning?

File: be/src/formats/orc/orc_chunk_reader.cpp (hunk 1 above)
Comment (xiaoyong-z) [reply be/src/formats/orc/orc_chunk_reader.cpp:2342]: If get_bytes gets error, and return.
Then the nbytes is 0? At this time, DCHECK_EQ(nbytes, length) will not pass?

File: be/src/formats/orc/orc_chunk_reader.cpp (hunk 2 above)
Comment (dirtysalt) [reply be/src/formats/orc/orc_chunk_reader.cpp:2305]: no, we use size in 
```
bool ORCHdfsFileStream::canUseCacheBuffer(uint64_t offset, uint64_t length) {
    if ((_cache_buffer.size() != 0) && (offset >= _cache_offset) &&
        ((offset + length) <= (_cache_offset + _cache_buffer.size()))) {
        return true;
    }
    return false;
}
```

File: be/src/formats/orc/orc_chunk_reader.cpp (hunk 3 above)
Comment (dirtysalt) [reply be/src/formats/orc/orc_chunk_reader.cpp:2344]: ideally yes. but our API is not that: we fill `buf` not return pointer.

> void ORCHdfsFileStream::doRead(void* buf, uint64_t length, uint64_t offset, bool direct);

File: be/src/formats/orc/orc_chunk_reader.cpp (hunk 4 above)
Comment (dirtysalt) [reply be/src/formats/orc/orc_chunk_reader.cpp:2307]: No. read from current position. For example

```
        S0 
 --> S1 [100-200]
        S2 [200-300]
```
when we read s1, we will s1 & s3 in a single shot.

**Review by chaoyli — APPROVED** [review]

File: be/src/formats/orc/orc_chunk_reader.cpp (hunk 1 above)
Comment (dirtysalt) [reply be/src/formats/orc/orc_chunk_reader.cpp:2342]: You are right. Better to check `st` before.

Comment (dirtysalt) [reply be/src/formats/orc/orc_chunk_reader.cpp:2342]: I'll fix it in another PR.

**Review by meegoo — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by xiaoyong-z — APPROVED** [review]

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Comment by dirtysalt** [comment]
run starrocks_be_unittest

**Pull request merged**